DSharpPlus.Core/Rest/PollyRatelimitPolicy.cs

```diff
@@ -0,0 +1,152 @@
+            {
+                throw new InvalidOperationException("No valid cache provided.");
+            }
+
+            // some endpoints are exempt from the global limit. account for them.
+            bool subjectToGlobalLimit = true;
+
+            if (context.TryGetValue("subject-to-global-limit", out object globalLimitObject)
+                && globalLimitObject is bool globallyLimited)
+            {
+                subjectToGlobalLimit = globallyLimited;
+            }
+
+            // already make our request, await it later
+            ValueTask<bool> awaitableBucket = cache.TryGetAsync(endpoint, out RatelimitBucket bucket);
+
+            // apply global ratelimits
+            // todo: handle a missing global bucket
+            if (subjectToGlobalLimit && await cache.TryGetAsync(GlobalBucketName, out RatelimitBucket globalBucket))
+            {
+                if (globalBucket is null)
+                {
+                    throw new InvalidOperationException("No global ratelimit bucket could be found.");
+                }
+
+                // store the current instant for a second here
+                DateTimeOffset instant = DateTimeOffset.UtcNow;
+
+                // if the bucket is already expired, reset it. technically we could save a bit of time by goto-ing out of the
+                // if clause entirely if this check succeeds, since we are then guaranteed the request will be allowed by the global bucket.
+                if (globalBucket.ResetTime < instant)
+                {
+                    globalBucket.ResetBucket(instant + __one_second);
+                }
+
+                if (!globalBucket.AllowNextRequest())
+                {
+                    HttpResponseMessage response = new(HttpStatusCode.TooManyRequests);
+
+                    response.Headers.RetryAfter = new(globalBucket.ResetTime - instant);
+
+                    // allow users to figure out whether the 429 showing up in their debug windows was thrown by Discord or DSharpPlus
+                    response.Headers.Add("X-DSharpPlus-Preemptive-RateLimit", "global");
+
+                    // pretend this was a discord api call and return the response.
+                    // the PollyRetryPolicy will see this 429 and retry the request later.
+                    return response;
+                }
+
+                await cache.CacheAsync(GlobalBucketName, globalBucket);
+            }
+
+            // if we don't already have this bucket cached, set the hash to our endpoint for now. 
+            // this will later be used to identify a newly called bucket.
+            string hash = __endpoint_hash_mapping.TryGetValue(endpoint, out string? nullableBucketHash)
+                ? nullableBucketHash
+                : endpoint;
+
+            if (await awaitableBucket)
+            {
+                // if our bucket is null for some reason, consider the request denied to avoid accidental 429s.
```

Comment:
`awaitableBucket` is always going to be false here, so this is never reached.

Reply:
why would it always be false?